rest_tools/client/client.py

```diff
@@ -46,13 +75,50 @@ def __init__(
         address: str,
         token: Optional[Union[str, bytes, Callable[[], Union[str, bytes]]]] = None,
         timeout: float = 60.0,
-        retries: int = 10,
+        retries: Union[int, CalcRetryFromBackoffMax, CalcRetryFromWaittimeMax] = 10,
+        backoff_factor: float = 0.3,
         logger: Optional[logging.Logger] = None,
-        **kwargs: Any
+        **kwargs: Any,
     ) -> None:
         self.address = address
+
         self.timeout = timeout
-        self.retries = retries
+        if self.timeout < 0.0:
+            raise ValueError(f"timeout must be positive: {self.timeout}")
+
+        self.backoff_factor = backoff_factor
+        if self.backoff_factor < 0.0:
+            raise ValueError(f"backoff_factor must be positive: {self.backoff_factor}")
+
+        if isinstance(retries, CalcRetryFromBackoffMax):
+            #                        backoff_last  =  backoff_factor * 2^retries
+            #       backoff_last / backoff_factor  =  2^retries
+            # lg( backoff_last / backoff_factor )  =  retries
+            self.retries = int(math.log2(retries.backoff_max / self.backoff_factor))
+        elif isinstance(retries, CalcRetryFromWaittimeMax):
+            # sum{n=1,k}(T + 2^n * B) + T  =  ( T*k + B*2^(k+1) - 2*B ) + T
+            # not easily solvable (for k) in a closed form
+            for k in range(1, MAX_RETRIES + 1):
+                if retries.waittime_max > (
+                    (self.timeout * k)
+                    + (self.backoff_factor * 2 ** (k + 1))
+                    - (2 * self.backoff_factor)
+                    + self.timeout
+                ):
+                    self.retries = k - 1  # get prev since it was <= waittime_max
+                    break
+            else:
+                self.retries = MAX_RETRIES + 1  # -> ValueError below
```

Comment:
Should this be `MAX_RETRIES`?  the loop didn't test that +1 was below the max time.  And it would break the validation below.

Reply:
yeah, it was asking for an off by one error. I refactored that chunk

Reply:
Do you want to hit the validation error if you go through the end of the loop?  You'll still hit that with the refactor.

Reply:
Can you write a test where it hits this case, to show you handle it the way you want?